apiary-ranger-metastore-plugin/src/main/java/com/expedia/apiary/extensions/rangerauth/listener/ApiaryRangerAuthPreEventListener.java

```diff
@@ -0,0 +1,230 @@
+
+import org.apache.hadoop.hive.metastore.events.PreEventContext;
+import org.apache.hadoop.hive.metastore.events.PreCreateTableEvent;
+import org.apache.hadoop.hive.metastore.events.PreDropTableEvent;
+import org.apache.hadoop.hive.metastore.events.PreAlterTableEvent;
+import org.apache.hadoop.hive.metastore.events.PreReadTableEvent;
+import org.apache.hadoop.hive.metastore.events.PreAddPartitionEvent;
+import org.apache.hadoop.hive.metastore.events.PreDropPartitionEvent;
+import org.apache.hadoop.hive.metastore.events.PreAlterPartitionEvent;
+import org.apache.hadoop.hive.metastore.events.PreAddIndexEvent;
+import org.apache.hadoop.hive.metastore.events.PreDropIndexEvent;
+import org.apache.hadoop.hive.metastore.events.PreAlterIndexEvent;
+import org.apache.hadoop.hive.metastore.events.PreReadDatabaseEvent;
+import org.apache.hadoop.hive.metastore.events.PreCreateDatabaseEvent;
+import org.apache.hadoop.hive.metastore.events.PreDropDatabaseEvent;
+
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.apache.hadoop.hive.ql.security.authorization.plugin.HiveOperationType;
+
+import org.apache.ranger.plugin.service.RangerBasePlugin;
+import org.apache.ranger.plugin.audit.RangerDefaultAuditHandler;
+import org.apache.ranger.plugin.policyengine.RangerAccessResourceImpl;
+import org.apache.ranger.plugin.policyengine.RangerAccessRequestImpl;
+import org.apache.ranger.plugin.policyengine.RangerAccessResult;
+
+
+enum HiveAccessType { NONE, CREATE, ALTER, DROP, INDEX, LOCK, SELECT, UPDATE, USE, READ, WRITE, ALL, ADMIN };
+
+
+/**
+ * A MetaStorePreEventListener to authorize using ranger.
+ */
+
+public class ApiaryRangerAuthPreEventListener extends MetaStorePreEventListener {
+
+  private static final Logger log = LoggerFactory.getLogger(ApiaryRangerAuthPreEventListener.class);
+
+  private static RangerBasePlugin plugin = null;
+
+  public ApiaryRangerAuthPreEventListener(Configuration config) throws HiveException {
+    super(config);
+    plugin = new RangerBasePlugin("hive","metastore");
+    plugin.init();
+    plugin.setResultProcessor(new RangerDefaultAuditHandler());
+    log.debug(" ApiaryRangerAuthPreEventListener created ");
+  }
+
+  public ApiaryRangerAuthPreEventListener(Configuration config,RangerBasePlugin plugin) throws HiveException {
+    super(config);
+    this.plugin = plugin;
+    log.debug("ApiaryRangerAuthPreEventListener created");
+  }
+
+  @Override
+  public void onEvent(PreEventContext context) throws MetaException, NoSuchObjectException, InvalidOperationException {
+
+    String user = null;
+    Set<String> groups = null;
+
+    try{
+        UserGroupInformation ugi = Utils.getUGI();
+        user = ugi.getUserName();
+        groups = Sets.newHashSet(ugi.getGroupNames());
+    }
+    catch(Exception ex)
+    {
+        log.debug("unable to read username", ex);
+        throw new InvalidOperationException("unable to read username.");
```

Comment:
Generally we prefer throwing the cause in any wrapped exceptions, i.e. 

`throw new InvalidOperationException("unable to read username.", ex);`

If you do this then you also don't need to log it as the stack trace of `ex` will be shown wherever the `InvalidOperationException` ends up.